// react-server/target/client/logging/response.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SuperLogger = require('winston').Transport,
    RLS = require('../util/RequestLocalStorage').getNamespace();

// A subset of stats that are logged are not associated with requests
// or occur before the request context is initialized. Simply ignore
// those logs here.
var queue = function queue() {
	if (RLS.isActive()) {
		return RLS().queue || (RLS().queue = []);
	} else {
		return [];
	}
};

var pushToQueue = function pushToQueue(module, lastModuleToken, key, level, msg, meta) {
	if (RLS.isActive() && !!RLS().doLog) {
		var tuple = [module, msg, meta[key], lastModuleToken];
		queue().push(tuple);
	}
};

var ResponseLogger = function (_SuperLogger) {
	_inherits(ResponseLogger, _SuperLogger);

	function ResponseLogger(options) {
		_classCallCheck(this, ResponseLogger);

		var _this = _possibleConstructorReturn(this, (ResponseLogger.__proto__ || Object.getPrototypeOf(ResponseLogger)).call(this));

		_this.name = 'ResponseLogger';
		_this.level = options.level || 'debug';
		_this.module = options.name;
		_this.lastModuleToken = options.name.split('.').pop();
		return _this;
	}

	_createClass(ResponseLogger, [{
		key: 'log',
		value: function log(level, msg, meta, callback) {
			pushToQueue(this.module, this.lastModuleToken, this.key, level, msg, meta);
			// Yield to the next log transport.
			callback(null, true);
		}
	}]);

	return ResponseLogger;
}(SuperLogger);

var TimeResponseLogger = function (_ResponseLogger) {
	_inherits(TimeResponseLogger, _ResponseLogger);

	function TimeResponseLogger(options) {
		_classCallCheck(this, TimeResponseLogger);

		var _this2 = _possibleConstructorReturn(this, (TimeResponseLogger.__proto__ || Object.getPrototypeOf(TimeResponseLogger)).call(this, options));

		_this2.name = 'TimeResponseLogger';
		_this2.level = 'fast';
		_this2.key = 'ms';
		return _this2;
	}

	return TimeResponseLogger;
}(ResponseLogger);

var GaugeResponseLogger = function (_ResponseLogger2) {
	_inherits(GaugeResponseLogger, _ResponseLogger2);

	function GaugeResponseLogger(options) {
		_classCallCheck(this, GaugeResponseLogger);

		var _this3 = _possibleConstructorReturn(this, (GaugeResponseLogger.__proto__ || Object.getPrototypeOf(GaugeResponseLogger)).call(this, options));

		_this3.name = 'GaugeResponseLogger';
		_this3.level = 'ok';
		_this3.key = 'val';
		return _this3;
	}

	return GaugeResponseLogger;
}(ResponseLogger);

var getTransportForGroup = function getTransportForGroup(group, opts) {
	if (group === "time") {
		return new TimeResponseLogger(opts);
	} else if (group === "gauge") {
		return new GaugeResponseLogger(opts);
	} else {
		return new ResponseLogger(opts);
	}
};

var flushLogsToResponse = function flushLogsToResponse(res) {
	if (queue().length > 0) {
		res.write("<script>");
		res.write('window.reactServerLogs = ' + JSON.stringify(queue()) + ';\n');
		res.write("</script>");
	}
};

var setResponseLoggerPage = function setResponseLoggerPage(page) {
	if (RLS.isActive() && !!page) {
		RLS().doLog = page.getRequest().getQuery()._debug_output_logs;
	}
};
module.exports = { setResponseLoggerPage: setResponseLoggerPage, flushLogsToResponse: flushLogsToResponse, getTransportForGroup: getTransportForGroup, TimeResponseLogger: TimeResponseLogger, ResponseLogger: ResponseLogger };
