// react-server/target/client/ClientController.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react'),
    ReactDOM = require('react-dom'),
    logging = require('./logging'),
    RequestContext = require('./context/RequestContext'),
    RequestLocalStorage = require('./util/RequestLocalStorage'),
    DebugUtil = require('./util/DebugUtil'),
    Q = require('q'),
    cssHelper = require('./util/ClientCssHelper'),
    EventEmitter = require("events").EventEmitter,
    ClientRequest = require("./ClientRequest"),
    History = require('./components/History'),
    PageUtil = require("./util/PageUtil"),
    ReactServerAgent = require('./ReactServerAgent'),
    _require = require('./components/RootElement'),
    getRootElementAttributes = _require.getRootElementAttributes,
    _require2 = require('./constants'),
    PAGE_LINK_NODE_ID = _require2.PAGE_LINK_NODE_ID,
    PAGE_CONTAINER_NODE_ID = _require2.PAGE_CONTAINER_NODE_ID;


var _ = {
	forEach: require('lodash/forEach'),
	assign: require('lodash/assign')
};

var RLS = RequestLocalStorage.getNamespace();

var logger = logging.getLogger({ "name": "react-server.core.ClientController", "color": { "server": 63, "client": "rgb(42,42,212)" } });

// for dev tools
window.React = React;

var REACT_SERVER_DATA_ATTRIBUTE = "data-react-server-root-id";

/**
 * Set up a Q error handler to make sure that errors that bubble
 * up are logged via our logger. Note: This will affect all unhandled
 * Q promise rejections, not just the ones in this file.
 */
Q.onerror = function (err) {
	logger.error("Unhandled exception in Q promise", err);
};

function getHistoryStateFrame(request) {

	// Mark the frame as ours.
	// Stash the request opts that were used to navigate to this frame.
	return { reactServerFrame: request ? request.getOpts() : {} };
}

function getHistoryPathname() {
	return location.pathname + location.search + location.hash;
}

var ClientController = function (_EventEmitter) {
	_inherits(ClientController, _EventEmitter);

	function ClientController(_ref) {
		var routes = _ref.routes;

		_classCallCheck(this, ClientController);

		var _this = _possibleConstructorReturn(this, (ClientController.__proto__ || Object.getPrototypeOf(ClientController)).call(this));

		window.__reactServerTimingStart = window.performance ? window.performance.timing.navigationStart : undefined;

		var dehydratedState = window.__reactServerState;

		checkNotEmpty(dehydratedState, 'InitialContext');
		checkNotEmpty(dehydratedState, 'Config');

		RequestLocalStorage.startRequest();

		_this.config = buildConfig(dehydratedState.Config);

		if (routes.onClientConfigLoaded) {
			routes.onClientConfigLoaded.call(_this);
		}

		_this.context = buildContext(routes);
		_this.context.setDeviceType(dehydratedState.InitialContext.deviceType);
		ReactServerAgent.cache().rehydrate(dehydratedState.InitialContext['ReactServerAgent.cache']);
		_this.mountNode = document.getElementById('content');

		_this._setupNavigateListener();
		_this._setupArrivalHandlers();

		_this._previouslyRendered = false;
		_this._rootNodeDfds = [];
		_this._failDfd = Q.defer();

		// Log this after loglevel is set.
		logTimingData('wakeFromStart', window.__reactServerTimingStart);

		performanceMark('wake');

		// this is a proxy for when above the fold content gets painted (displayed) on the browser
		logTimingData('displayAboveTheFold.fromStart', window.__reactServerTimingStart, window.__displayAboveTheFold);
		return _this;
	}

	_createClass(ClientController, [{
		key: 'terminate',
		value: function terminate() {
			// We may not have set up any history stuff...
			if (!this._history) return;
			this._history.off(this._historyListener);
			this._historyListener = null;
			this._history = null;
		}
	}, {
		key: '_startRequest',
		value: function _startRequest(_ref2) {
			var request = _ref2.request,
			    type = _ref2.type;


			var t0 = type === History.events.PAGELOAD ? window.__reactServerTimingStart // Try to use navigation timing.
			: new Date(); // There's no naviagation.  We're it.

			var url = request.getUrl();
			var isPush = type === History.events.PUSHSTATE;

			// This is the navigator we're going to listen to regarding when navigation
			// is complete for timing purposes.
			var navigationTimingAuthority = this.context.navigator;

			this._reuseDom = request.getReuseDom();

			if (this._previouslyRendered) {

				// This is a secondary request (client transition)
				// within a session, so we'll get a fresh
				// RequestLocalStorage container.
				RequestLocalStorage.startRequest();

				// If we're not going to reuse the DOM, let's
				// clean up right away to blank the screen.
				if (!this._reuseDom) {
					this._cleanupPreviousRender();
				}

				// we need to re-register the request context
				// as a RequestLocal.
				this.context.registerRequestLocal();
			}

			// If this is a History.events.PUSHSTATE navigation,
			// and we have control of the navigation bar we should
			// change the URL in the location bar before rendering.
			//
			// Note that for browsers that do not have pushState,
			// this will result in a window.location change and
			// full browser load.
			//
			if (this._history) {

				if (isPush) {

					// Sorry folks.  If we need to do a client
					// transition, then we're going to clobber
					// your state.  You must be able to render
					// from URL, anyway, so if you're set up right
					// it won't affect user experience.  It means,
					// though, that there exists a navigation path
					// to an extraneous full-page rebuild.
					// Such is life.
					if (
					// Don't replace state unless we've
					// got a real history API.
					this._history.canClientNavigate() && !(history.state || {}).reactServerFrame) {
						this._history.replaceState(getHistoryStateFrame(), null, getHistoryPathname());
					}

					this._setHistoryRequestOpts({
						// If we're reusing the DOM on the way
						// forward, then we can also reuse on
						// the way back.
						reuseDom: request.getReuseDom()
					});

					this._history.pushState(getHistoryStateFrame(request), null, url);
				} else if (type === History.events.PAGELOAD) {

					// This _seems_ redundant with the
					// `replaceState` above, but keep in mind that
					// an initial `pushState` might not be a
					// client transition.  It could be a
					// non-`react-server` use of the history API.
					//
					// This also _replaces_ state with the request
					// URL, which handles client-side redirects.
					this._history.replaceState(getHistoryStateFrame(request), null, url);

					this._setHistoryRequestOpts({

						// If we wind up back here without
						// first client-transitioning away
						// then presumably we're still on the
						// same page that just had some
						// history maniptulation outside of
						// `react-server`.  In that case we're
						// ourselves and we should be able to
						// re-use the DOM.  Maybe
						// presumptuous, but a nicer
						// experience than clobbering.
						reuseDom: true
					});
				}
			}

			// logging how long the request took in a variety of ways:
			// - Request type (pageload, pushstate, popstate)
			// - Request options (reuseDom, bundleData, etc)
			navigationTimingAuthority.once('loadComplete', function () {
				var bas = 'handleRequest';
				var typ = 'type.' + (type || 'PAGELOAD');
				logTimingData(bas + '.all', t0);
				logTimingData(bas + '.' + typ + '.all', t0);
				_.forEach(request.getOpts(), function (val, key) {
					if (val) {
						var opt = 'opt.' + key;
						logTimingData(bas + '.' + opt, t0);
						logTimingData(bas + '.' + typ + '.' + opt, t0);
					}
				});
			});

			this._lastState = history.state;
		}

		// Update the request options for the _current_ history navigation
		// state frame prior to pushing a new frame.

	}, {
		key: '_setHistoryRequestOpts',
		value: function _setHistoryRequestOpts(opts) {

			// If we don't have a real history API then we don't want to
			// mess with state since it results in full navigation.
			if (!this._history.canClientNavigate()) return;

			var state = _.assign({}, history.state);
			state.reactServerFrame = _.assign(state.reactServerFrame || {}, opts);
			this._history.replaceState(state, null, getHistoryPathname());
		}
	}, {
		key: '_setupNavigateListener',
		value: function _setupNavigateListener() {
			var _this2 = this;

			var context = this.context;

			context.onNavigateStart(this._startRequest.bind(this));

			/**
    * type is one of
    *    History.events.PUSHSTATE: user clicked something to go forward but browser didn't do a
    * full page load
    *    History.events.POPSTATE: user clicked back button but browser didn't do a full page load
    *    History.events.PAGELOAD: full browser page load, not using History API.
    */
			context.onNavigate(function (err, page) {
				logger.debug('Executing navigate action');

				_this2._handleDebugParams(page);

				if (err) {
					// redirects are sent as errors, so let's handle it if that's the case.
					if (err.status && (err.status === 301 || err.status === 302)) {
						if (!err.redirectUrl) {
							console.error("A redirect status was sent without a corresponding redirect redirectUrl.", err);
						} else {
							setTimeout(function () {
								// we're about to load the _next_ page, so we should mark the
								// redirect navigation finished
								context.navigator.finishRoute();
								context.navigate(new ClientRequest(err.redirectUrl));
							}, 0);
						}
					} else {
						logger.error("onNavigate error", err);
					}
					return;
				}

				var routeName = context.navigator.getCurrentRoute().name;

				if (!_this2._previouslyRendered) {
					cssHelper.registerPageLoad(routeName);
				} else {
					// render the document title.
					_this2._renderTitle(page);

					// render the base tag.
					_this2._renderBase(page);

					_this2._renderMetaTags(page);

					_this2._renderLinkTags(page);
				}

				cssHelper.ensureCss(routeName, page).then(function () {
					return page.getBodyClasses().then(function (classes) {
						classes.push('route-' + routeName);
						document.body.className = classes.join(' ');
					});
				}).then(function () {
					return _this2._render(page);
				}).catch(function (err) {
					logger.error("Error during client transition render", err);
				}).then(function () {

					// We're responsible for letting the navigator
					// know when we're more or less done stomping
					// around in the current request context
					// setting things up.
					//
					// We can't _guarantee_ that pages/middleware
					// haven't set timers to mess with things in
					// the future, so we need to wait a bit before
					// letting the navigator yank our context if
					// an immediate subsequent navigation is
					// scheduled.
					//
					// I don't like this magic delay here, but it
					// gives us a better shot at falling after
					// things like lazy load images do their
					// post-render wire-up.
					//
					// Anything that the current page does in the
					// request context _after_ this timeout has
					// elapsed and we've started a subsequent
					// navigation is pure corruption. :p
					//
					setTimeout(function () {
						return context.navigator.finishRoute();
					}, 200);
				}).done();
			});
		}
	}, {
		key: '_handleDebugParams',
		value: function _handleDebugParams(page) {
			if (!page) return;

			DebugUtil.setRequest(page.getRequest());

			var params = DebugUtil.getAllDebugValues();

			// Allow adjustment of log levels.
			_.forEach({
				_react_server_log_level: 'main',
				_react_server_log_level_main: 'main',
				_react_server_log_level_time: 'time',
				_react_server_log_level_gauge: 'gauge'
			}, function (type, param) {
				if (params[param]) {
					logging.setLevel(type, params[param]);
				}
			});
		}
	}, {
		key: '_renderTitle',
		value: function _renderTitle(page) {
			var _this3 = this;

			page.getTitle().then(function (newTitle) {
				if (newTitle && newTitle !== document.title) {
					document.title = newTitle;
				}

				// This is the earliest we have everything we need for
				// an analytics pageview event.
				_this3.emit("pageview", {
					page: getHistoryPathname(),
					title: newTitle
				});
			}).catch(function (err) {
				logger.error("Error while setting the document title", err);
			}).done();
		}
	}, {
		key: '_renderBase',
		value: function _renderBase(page) {
			page.getBase().then(function (base) {
				var currentBaseTag = document.head.querySelector("head base");
				if (base === null) {
					// get rid of the current base tag.
					if (currentBaseTag) currentBaseTag.parentNode.removeChild(currentBaseTag);
				} else {
					// we need a base tag. add one if it's not there yet.
					if (!currentBaseTag) {
						currentBaseTag = document.createElement("base");
						document.head.appendChild(currentBaseTag);
					}
					if (base.href) currentBaseTag.href = base.href;
					if (base.target) currentBaseTag.target = base.target;
				}
			}).catch(function (err) {
				logger.error("Error rendering <base>", err);
			}).done();
		}
	}, {
		key: '_renderMetaTags',
		value: function _renderMetaTags(page) {
			// first, remove all the current meta tags.
			var currentMetaTags = document.head.querySelectorAll("meta");
			for (var i = 0; i < currentMetaTags.length; i++) {
				currentMetaTags[i].parentNode.removeChild(currentMetaTags[i]);
			}

			// now add all the meta tags for the new page.
			page.getMetaTags().forEach(function (metaTagPromise) {
				metaTagPromise.then(function (metaTag) {
					var parent = document.head;
					if (metaTag.noscript) {
						var noscript = document.createElement("noscript");
						parent.appendChild(noscript);
						parent = noscript;
					}

					var meta = document.createElement("meta");
					["name", "httpEquiv", "charset", "content"].forEach(function (name) {
						if (metaTag[name]) meta[name] = metaTag[name];
					});

					parent.appendChild(meta);
				}).catch(function (err) {
					logger.error("Error rendering meta tags", err);
				}).done();
			});
		}
	}, {
		key: '_renderLinkTags',
		value: function _renderLinkTags(page) {

			// First, remove all the current link tags.
			;[].slice.call(document.head.querySelectorAll('link[' + PAGE_LINK_NODE_ID + ']')).forEach(function (tag) {
				return tag.parentNode.removeChild(tag);
			});

			// Then add all the link tags for the new page.
			page.getLinkTags().forEach(function (promise) {
				return promise.then(PageUtil.makeArray).then(function (tags) {
					return tags.forEach(function (tag) {
						document.head.appendChild([document.createElement('link'), PAGE_LINK_NODE_ID].concat(Object.keys(tag)).reduce(function (link, attr) {
							return link.setAttribute(attr, tag[attr] || ''), link;
						}));
					});
				}).catch(function (err) {
					return logger.error("Error rendering link tags", err);
				}).done();
			});
		}
	}, {
		key: '_render',
		value: function _render(page) {
			var _this4 = this;

			var tStart = window.__reactServerTimingStart;
			var t0 = new Date();
			var retval = Q.defer();

			logger.debug('React Rendering');

			// We keep track of the _total_ time we spent rendering during
			// each request so we can keep track of that overhead.
			var totalRenderTime = 0;

			// These resolve with React elements when their data
			// dependencies are fulfilled.
			var elementPromises = PageUtil.standardizeElements(page.getElements());
			var timeoutDfd = [];
			var elementPromisesOr = elementPromises.map(function (promise, index) {
				var orPromise = Q.defer();
				timeoutDfd[index] = Q.defer();

				promise.then(orPromise.resolve);
				promise.catch(orPromise.reject);
				timeoutDfd[index].promise.catch(orPromise.reject);

				return orPromise.promise;
			});

			// These resolve with DOM mount points for the elements.
			//
			// Our behavior is different here for the _first_ render vs
			// during a client transition.
			var rootNodePromises;
			if (this._previouslyRendered) {

				// On a client transition we've just blown away all of
				// our mount points from the previous page, and we'll
				// create a fresh set.  We'll defer creating them
				// until we've actually got our elements, since some
				// items in the elements array may be container
				// control.
				rootNodePromises = elementPromises.map(function () {
					return Q();
				});
			} else {

				// On our _first_ render we want to mount to the DOM
				// nodes produced during the _server-side_ render.
				//
				// We're awake and doing our thing while these
				// server-rendered elements are streaming down, so we
				// need to wait to render a given element until its
				// mount point arrives.
				//
				// The server will tell us when each mount point is
				// ready by calling `nodeArrival`, which triggers
				// resolution of the corresponding `rootNodePromise`.
				elementPromises.forEach(function (promise, index) {
					_this4._ensureRootNodeDfd(index);
				});
				rootNodePromises = this._rootNodeDfds.map(function (dfd) {
					return dfd.promise;
				});
			}

			var mountNode = this.mountNode;

			// These are only used if we're going to try to re-use the
			// existing DOM structure.
			var oldRootElement, oldRootContainer;

			// Once we've got an element and a root DOM node to mount it
			// in we can finally render.
			var renderElement = function renderElement(element, root, index) {

				// During client transitions we create our root
				// elements as we go.
				if (!root && _this4._previouslyRendered) {

					// If the _previous_ render had elements that
					// we can re-use we'll render into them.
					//
					// DOM re-use is currently opt-in.
					//
					if (_this4._reuseDom) {
						oldRootElement = document.querySelector('div[' + REACT_SERVER_DATA_ATTRIBUTE + '="' + index + '"]');
						oldRootContainer = document.querySelector('div[' + PAGE_CONTAINER_NODE_ID + '="' + index + '"]');
					}

					// The current strategy for re-use is: So long
					// as the _shape_ of the root structure is the
					// same, we'll re-use.  Once the new page's
					// shape diverges, we'll blow away the
					// remaining elements left over from the
					// previous page and create everything for the
					// new page as we go.
					//
					if (_this4._reuseDom && element.containerOpen && oldRootContainer) {
						mountNode = oldRootContainer;
						_this4._updateContainerNodeAttributes(mountNode, element.containerOpen);
					} else if (_this4._reuseDom && element.containerClose && !oldRootContainer && !oldRootElement) {
						mountNode = mountNode.parentNode;
					} else if (_this4._reuseDom && oldRootElement) {
						root = oldRootElement;
					} else {
						_this4._cleanupPreviousRender(index);
						if (element.containerOpen) {

							// If we're opening a container that's
							// our new mountNode.
							mountNode = _this4._createContainerNode(mountNode, element.containerOpen, index);
						} else if (element.containerClose) {

							// If we're closing a container its
							// parent is once again our mountNode.
							mountNode = mountNode.parentNode;
						} else if (!element.isTheFold) {

							// Need a new root element in our
							// current mountNode.
							root = _this4._createReactServerRootNode(mountNode, index);
						}
					}
				}

				if (element.containerOpen || element.containerClose) {
					return; // Nothing left to do.
				} else if (element.isTheFold) {
					if (!_this4._previouslyRendered) {
						logTimingData('renderAboveTheFold.fromStart', tStart);
						logTimingData('renderAboveTheFold.individual', 0, totalRenderTime);
						logTimingData('renderAboveTheFold.elementCount', 0, index + 1);

						performanceMark('renderAboveTheFold');
					}
					return; // Again, this isn't a real root element.
				}

				var name = PageUtil.getElementDisplayName(element),
				    timer = logger.timer('renderElement.individual.' + name);

				element = React.cloneElement(element, { context: _this4.context });
				var renderFunc = ReactDOM.hydrate || ReactDOM.render;
				renderFunc(element, root);

				_.forEach(getRootElementAttributes(element), function (v, k) {
					return root.setAttribute(k, v);
				});

				totalRenderTime += timer.stop();

				if (!_this4._previouslyRendered) {
					var tDisplay = root.getAttribute('data-react-server-timing-offset');
					logTimingData('displayElement.fromStart.' + name, 0, +tDisplay);
					logTimingData('renderElement.fromStart.' + name, tStart);
				}
			};

			var renderOne = function renderOne(promise, index) {
				return promise.then(function (element) {
					return rootNodePromises[index].then(function (root) {
						return renderElement(element, root, index);
					}).catch(function (e) {
						// The only case where this should evaluate to false is
						// when `element` is a containerClose/containerOpen object
						var componentType = typeof element.type === 'function' ? element.props.children.type.name : 'element';
						logger.error('Error with element ' + componentType + '\'s lifecycle methods at index ' + index, e);
					});
				}).catch(function (e) {
					return logger.error('Error with element promise ' + index, e);
				});
			};

			if (this._previouslyRendered) {

				// On client transitions the root structure is laid out using a
				// state machine that requires us to render in order.
				elementPromisesOr.reduce(function (chain, promise, index) {
					return chain.then(function () {
						return renderOne(promise, index);
					});
				}, Q()).then(retval.resolve);
			} else {

				// On the first render we can go out of order because the server
				// has already laid out the root structure for us.
				Q.all(elementPromisesOr.map(renderOne)).then(retval.resolve);

				// Look out for a failsafe timeout from the server on our
				// first render.
				this._failDfd.promise.then(function () {
					elementPromises.forEach(function (promise, index) {
						//Reject any elements that have failed to render
						if (promise.isPending()) {
							timeoutDfd[index].reject('Error with element ' + index + ', it failed to render within timeout time');
						}
					});
				});
			}

			return retval.promise.then(function () {

				if (_this4._reuseDom) {

					// Clean up any dangling nodes if the previous page had more
					// than we do.
					_this4._cleanupPreviousRender(elementPromises.length);
				}

				// This first one is just for historical continuity.
				logTimingData('render', t0);

				// These are more interesting.
				logTimingData('renderCPUTime', 0, totalRenderTime);

				// Don't track this on client transitions.
				if (!_this4._previouslyRendered) {
					logTimingData('renderFromStart', tStart);

					performanceMark('renderComplete');
				}

				// Some things are just different on our first pass.
				_this4._previouslyRendered = true;

				_this4.emit('render');
			});
		}

		/**
   * Cleans up a previous React render in the document. Unmounts all the components and destoys the mounting
   * DOM node(s) that were created.
   */

	}, {
		key: '_cleanupPreviousRender',
		value: function _cleanupPreviousRender(index) {

			index = index || 0; // Default to everything.

			if (this._previouslyRendered && !RLS().haveCleanedPreviousRender) {

				// Only need to do this once per request.
				RLS().haveCleanedPreviousRender = true;

				logger.debug("Removing previous page's React components");

				[].slice.call(document.querySelectorAll('div[' + REACT_SERVER_DATA_ATTRIBUTE + ']')).forEach(function (root, i) {
					if (i >= index) {
						// Since this node has a "data-react-server-root-id"
						// attribute, we can assume that we created it
						// and should destroy it. Destruction means
						// first unmounting from React and then
						// destroying the DOM node.
						ReactDOM.unmountComponentAtNode(root);
						root.parentNode.removeChild(root);
					}
				});

				[].slice.call(document.querySelectorAll('div[' + PAGE_CONTAINER_NODE_ID + ']')).forEach(function (root, i) {
					if (i >= index) {
						// Gotta get rid of our containers,
						// too.  Need to do this _after_
						// killing the elements, since they
						// might live within these containers.
						root.parentNode.removeChild(root);
					}
				});
			}
		}

		/**
   * This method creates a new div to render a ReactElement in to at the end of the mount node.
   */

	}, {
		key: '_createReactServerRootNode',
		value: function _createReactServerRootNode(mountNode, index) {
			var root = document.createElement("div");
			root.setAttribute(REACT_SERVER_DATA_ATTRIBUTE, index);
			mountNode.appendChild(root);
			return root;
		}
	}, {
		key: '_createContainerNode',
		value: function _createContainerNode(mountNode, attrs, i) {
			var node = document.createElement("div");
			node.setAttribute(PAGE_CONTAINER_NODE_ID, i);
			_.forEach(attrs, function (v, k) {
				return node.setAttribute(k, v);
			});
			mountNode.appendChild(node);
			return node;
		}
	}, {
		key: '_updateContainerNodeAttributes',
		value: function _updateContainerNodeAttributes(node, attrs) {
			_.forEach(attrs, function (v, k) {
				return node.setAttribute(k, v);
			});
		}
	}, {
		key: 'init',
		value: function init() {
			var _this5 = this;

			var unloadHandler = function unloadHandler() {
				_this5.terminate();
			};

			if (window && window.addEventListener) {
				window.addEventListener("unload", unloadHandler);
			} else if (window && window.attachEvent) {
				window.attachEvent("onunload", unloadHandler);
			}

			this._initializeHistoryListener(this.context);

			// If this is a _refresh_ there may be some request options
			// stashed in the history navigation stack frame we're sitting
			// on.
			var state = this._history.canClientNavigate() && history.state;

			this._navigateWithHistoryState({
				state: state,
				path: getHistoryPathname(),
				type: History.events.PAGELOAD
			});
		}
	}, {
		key: '_navigateWithHistoryState',
		value: function _navigateWithHistoryState(_ref3) {
			var path = _ref3.path,
			    state = _ref3.state,
			    type = _ref3.type,
			    check = _ref3.check;

			var opts = (state || {}).reactServerFrame;

			if (check && !opts) return; // Not our frame.

			this.context.navigate(new ClientRequest(path, opts || {}), type);
		}

		/**
   * Initializes us to listen to back button events. When the user presses the back button, the history
   * listener will be called and cause a navigate() event.
   */

	}, {
		key: '_initializeHistoryListener',
		value: function _initializeHistoryListener() {
			var _this6 = this;

			this._historyListener = function (_ref4) {
				var state = _ref4.state;

				_this6._navigateWithHistoryState({
					state: state,
					path: _this6._history.getPath(),
					type: History.events.POPSTATE, // Forward/back.
					check: true // Only navigate if frame is ours.
				});
			};

			this._history = new History();
			var init = function init() {
				return _this6._history.on(_this6._historyListener);
			};

			// Need to go _after_ 'load' callbacks complete.
			// Safari fires a 'popstate' on load (RED-67600).
			// https://developer.mozilla.org/en-US/docs/Web/Events/popstate
			if (document.readyState === 'complete') {
				init();
			} else {
				window.addEventListener('load', function () {
					return setTimeout(init, 0);
				});
			}
		}
	}, {
		key: '_setupArrivalHandlers',
		value: function _setupArrivalHandlers() {
			// used by <script> callbacks to register data sent down on the
			// initial connection after initial render
			window.__reactServerClientController = this;
		}
	}, {
		key: '_ensureRootNodeDfd',
		value: function _ensureRootNodeDfd(index) {
			if (!this._rootNodeDfds[index]) {
				this._rootNodeDfds[index] = Q.defer();
			}
			return this._rootNodeDfds[index];
		}
	}, {
		key: 'dataArrival',
		value: function dataArrival(url, dehydratedEntry) {
			ReactServerAgent.cache().lateArrival(url, dehydratedEntry);
		}
	}, {
		key: 'nodeArrival',
		value: function nodeArrival(startIndex, endIndex) {

			// The server has just let us know that a pre-rendered root
			// element has arrived.  We'll grab a reference to its DOM
			// node and un-block client-side rendering of the element that
			// we're going to mount into it.
			for (var i = startIndex; i <= endIndex; i++) {
				this._ensureRootNodeDfd(i).resolve(this.mountNode.querySelector('div[' + REACT_SERVER_DATA_ATTRIBUTE + '="' + i + '"]'));
			}
		}
	}, {
		key: 'failArrival',
		value: function failArrival() {
			this._failDfd.resolve();
		}
	}]);

	return ClientController;
}(EventEmitter);

function checkNotEmpty(state, key) {
	if (typeof state[key] === 'undefined') {
		var msg = key + ' not defined in dehydrated state';
		logger.error(msg);
		throw new Error(msg);
	}
}

function buildConfig(dehydratedConfig) {
	// rehydrate the config object
	var config = require("./config")();
	config.rehydrate(dehydratedConfig);
	return config;
}

function buildContext(routes) {
	var context = new RequestContext.Builder().setRoutes(routes).create();

	return context;
}

// Create a "user timing" in WebPageTest.
function performanceMark(name) {
	if (window.performance && performance.mark) {
		performance.mark('react-server.' + name);
	}
}

function logTimingData(bucket, start) {
	var end = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : new Date();

	if (start === undefined) {
		//don't send timing data if start timing is undefined
		return;
	}

	logger.time(bucket, end - start);
}

module.exports = ClientController;
