// react-server/target/client/ClientRequest.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var cookie = require("cookie"),
    decode = require("querystring/decode");

/**
 * This class implements the react-server request API for client-side transitions.
 */

var ClientRequest = function () {
	function ClientRequest(url) {
		var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
		    bundleData = _ref.bundleData,
		    reuseDom = _ref.reuseDom;

		_classCallCheck(this, ClientRequest);

		this._opts = {
			bundleData: bundleData,
			reuseDom: reuseDom

			// Chop off the fragment identifier from the url i.e everything from the # to the end of the url
			// if it exists to make this consistent with ExpressServerRequest.
		};var match = url.match(/([^#]*)/);
		if (match === null || !match[1]) {
			this._url = url;
		} else {
			this._url = match[1];
		}
	}

	_createClass(ClientRequest, [{
		key: "setRoute",
		value: function setRoute(route) {
			this._route = route;
		}
	}, {
		key: "getUrl",
		value: function getUrl() {
			return this._url;
		}
	}, {
		key: "getOpts",
		value: function getOpts() {
			return this._opts;
		}
	}, {
		key: "getReuseDom",
		value: function getReuseDom() {
			return this._opts.reuseDom;
		}
	}, {
		key: "getBundleData",
		value: function getBundleData() {
			return this._opts.bundleData;
		}
	}, {
		key: "getQuery",
		value: function getQuery() {
			// Grab fragment after first "?"
			var match = this._url.match(/\?(.*)/);

			if (match === null || !match[1]) {
				return {};
			}

			return decode(match[1]);
		}
	}, {
		key: "getHostname",
		value: function getHostname() {
			var hostname = null;
			if (typeof window.location.hostname === "string") {
				hostname = window.location.hostname;
			}

			return hostname;
		}
	}, {
		key: "getProtocol",
		value: function getProtocol() {
			var proto = null;
			if (typeof window.location.protocol === "string") {
				proto = window.location.protocol.replace(/:/g, '');
			}

			return proto;
		}
	}, {
		key: "getSecure",
		value: function getSecure() {
			return 'https' === this.getProtocol();
		}
	}, {
		key: "getRouteParams",
		value: function getRouteParams() {
			return this._route.params;
		}
	}, {
		key: "getMethod",
		value: function getMethod() {
			// I believe that client-side transitions should always be
			// HTTP GETs -sra.
			return "get";
		}
	}, {
		key: "getRouteName",
		value: function getRouteName() {
			return this._route.name;
		}
		/*eslint-disable no-unused-vars */

	}, {
		key: "getHttpHeader",
		value: function getHttpHeader(name) {
			// this is a no-op; there are no HTTP headers on the client-side.
		}
	}, {
		key: "getHttpHeaders",
		value: function getHttpHeaders() {}
		// this is a no-op; there are no HTTP headers on the client-side.

		/*eslint-enable no-unused-vars */

	}, {
		key: "getCookie",
		value: function getCookie(name) {
			return this.getCookies()[name];
		}
	}, {
		key: "getCookies",
		value: function getCookies() {

			return cookie.parse(document.cookie);
		}
	}, {
		key: "getBody",
		value: function getBody() {
			console.error("ClientRequest.getBody not implemented.");
		}
	}]);

	return ClientRequest;
}();

module.exports = ClientRequest;
