// react-server/target/client/ExpressServerRequest.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * This class wraps a Request object from the Express Server and provides the react-server Request
 * API.
 */
var ExpressServerRequest = function () {
	function ExpressServerRequest(expressServerRequest) {
		_classCallCheck(this, ExpressServerRequest);

		this._wrappedRequest = expressServerRequest;
	}

	_createClass(ExpressServerRequest, [{
		key: "setRoute",
		value: function setRoute(route) {
			this._route = route;
		}
	}, {
		key: "getUrl",
		value: function getUrl() {
			return this._wrappedRequest.url;
		}
	}, {
		key: "getQuery",
		value: function getQuery() {
			return this._wrappedRequest.query;
		}
	}, {
		key: "getHostname",
		value: function getHostname() {
			return this._wrappedRequest.hostname;
		}
	}, {
		key: "getProtocol",
		value: function getProtocol() {
			return this._wrappedRequest.protocol;
		}
	}, {
		key: "getSecure",
		value: function getSecure() {
			return this._wrappedRequest.secure;
		}
	}, {
		key: "getRouteParams",
		value: function getRouteParams() {
			return this._route.params;
		}
	}, {
		key: "getMethod",
		value: function getMethod() {
			return this._wrappedRequest.method;
		}
	}, {
		key: "getHttpHeader",
		value: function getHttpHeader(name) {
			return this._wrappedRequest.get(name);
		}
	}, {
		key: "getRouteName",
		value: function getRouteName() {
			return this._route.name;
		}
	}, {
		key: "getHttpHeaders",
		value: function getHttpHeaders() {
			return this._wrappedRequest.headers;
		}
	}, {
		key: "getCookie",
		value: function getCookie(name) {
			return this.getCookies()[name];
		}
	}, {
		key: "getCookies",
		value: function getCookies() {
			return this._wrappedRequest.cookies;
		}
	}, {
		key: "getBody",
		value: function getBody() {
			return this._wrappedRequest.body;
		}
	}, {
		key: "getBundleData",
		value: function getBundleData() {
			return false; // Not on the server.
		}
	}]);

	return ExpressServerRequest;
}();

module.exports = ExpressServerRequest;
