// react-server/target/client/ReactServerAgent.js
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var RLS = require('./util/RequestLocalStorage').getNamespace(),
    Cache = require("./ReactServerAgent/Cache"),
    Request = require("./ReactServerAgent/Request"),
    Plugins = require("./ReactServerAgent/Plugins");

var DATA_BUNDLE_PARAMETER = '_react_server_data_bundle';
var DATA_BUNDLE_OPTS = _defineProperty({}, DATA_BUNDLE_PARAMETER, 1);

var API = {

	DATA_BUNDLE_PARAMETER: DATA_BUNDLE_PARAMETER,

	get: function get(url, data) {
		var req = new Request('GET', url, API.cache());
		if (data) req.query(data);
		return req;
	},
	head: function head(url, data) {
		var req = new Request('HEAD', url, API.cache());
		if (data) req.query(data);
		return req;
	},
	del: function del(url, data) {
		var req = new Request('DELETE', url, API.cache());
		if (data) req.send(data);
		return req;
	},
	patch: function patch(url, data) {
		var req = new Request('PATCH', url, API.cache());
		if (data) req.send(data);
		return req;
	},
	post: function post(url, data) {
		var req = new Request('POST', url, API.cache());
		if (data) req.send(data);
		return req;
	},
	put: function put(url, data) {
		var req = new Request('PUT', url, API.cache());
		if (data) req.send(data);
		return req;
	},


	/**
  * Exposes the ReactServerAgent request data cache from RequestLocalStorage.
  */
	cache: function cache() {
		var cache = RLS().cache;
		if (!cache) {
			cache = RLS().cache = new Cache();
		}
		return cache;
	},
	_clearCache: function _clearCache() {
		delete RLS().cache;
	},


	/**
  * Adds a plugin function that can be used to modify the Request
  * object before the request is actually
  * triggered.
  *
  * The callback function will take the Request instance as a parameter:
  * ```
  * var defaultHeaders = { ... };
  * ReactServerAgent.plugRequest(function (request) {
  *     // e.g.
  *     request.set(defaultHeaders)
  * })
  * ```
  */
	plugRequest: function plugRequest(pluginFunc) {
		Plugins.forRequest().add(pluginFunc);
	},


	/**
  * Adds a plugin function that can be used to modify the response
  * object before it is passed the caller's callback.
  *
  * The callback function will take err and response as parameters
  * (like the callback to `end()`), and the request as well:
  * ```
  * ReactServerAgent.plugResponse(function (err, response, request) {
  *     // e.g.
  *     console.log("Response received!", res.body);
  *     res.wasLogged = true; // or whatever
  * })
  * ```
  */
	plugResponse: function plugResponse(pluginFunc) {
		Plugins.forResponse().add(pluginFunc);
	},
	_fetchDataBundle: function _fetchDataBundle(url) {
		return API.get(url, DATA_BUNDLE_OPTS).then(function (data) {
			return JSON.stringify(data.body);
		});
	},
	_rehydrateDataBundle: function _rehydrateDataBundle(data) {
		API.cache().rehydrate(JSON.parse(data));
	}
};

module.exports = API;
