// react-server/target/server/components/Link.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react'),
    PropTypes = require('prop-types'),
    navigateTo = require("../util/navigateTo");

var Link = function (_React$Component) {
	_inherits(Link, _React$Component);

	_createClass(Link, null, [{
		key: 'displayName',
		get: function get() {
			return 'Link';
		}
	}, {
		key: 'propTypes',
		get: function get() {
			return {
				path: PropTypes.string,
				href: PropTypes.string,
				onClick: PropTypes.func,
				bundleData: PropTypes.bool,
				reuseDom: PropTypes.bool,
				className: PropTypes.string
			};
		}
	}, {
		key: 'defaultProps',
		get: function get() {
			return {
				bundleData: false,
				reuseDom: false
			};
		}
	}]);

	function Link(props) {
		_classCallCheck(this, Link);

		var _this = _possibleConstructorReturn(this, (Link.__proto__ || Object.getPrototypeOf(Link)).call(this, props));

		_this._onClick = _this._onClick.bind(_this);
		return _this;
	}

	_createClass(Link, [{
		key: 'render',
		value: function render() {
			return React.createElement(
				'a',
				{ href: this.props.path || this.props.href, onClick: this._onClick, className: this.props.className },
				this.props.children
			);
		}
	}, {
		key: '_onClick',
		value: function _onClick(e) {

			// TODO: IE8-9 detection

			// TODO: if OSX && key.isMeta?
			if (!e.metaKey) {
				e.preventDefault();
				e.stopPropagation();
				var _props = this.props,
				    bundleData = _props.bundleData,
				    reuseDom = _props.reuseDom;

				navigateTo(this.props.path || this.props.href, {
					bundleData: bundleData,
					reuseDom: reuseDom
				});
				if (this.props.onClick) {
					this.props.onClick(e);
				}
			} else {
				// do normal browser navigate
			}
		}
	}]);

	return Link;
}(React.Component);

module.exports = Link;
