// react-server/target/server/util/DebugUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _RequestLocalStorage = require("./RequestLocalStorage");

var _RequestLocalStorage2 = _interopRequireDefault(_RequestLocalStorage);

var _forEach = require("lodash/forEach");

var _forEach2 = _interopRequireDefault(_forEach);

var _map = require("lodash/map");

var _map2 = _interopRequireDefault(_map);

var _assign = require("lodash/assign");

var _assign2 = _interopRequireDefault(_assign);

var _pick = require("lodash/pick");

var _pick2 = _interopRequireDefault(_pick);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RLS = _RequestLocalStorage2.default.getNamespace();

// This module provides methods that expose values from the query string.
//
// So, for example, if you pass `?_debug_render_timeout=1000` then you can
// call `DebugUtil.getRenderTimeout()` and receive `1000`.
//
// Additionally you may call `DebugUtil.getAllDebugValues()` to obtain an
// object with all debug parameters extracted from the query string.
//
var DEBUG_PARAMS = {
	getRenderTimeout: "_debug_render_timeout",
	getOutputLogs: "_debug_output_logs",
	getLab: "_debug_lab",
	getJsBelowTheFold: "_debug_js_below_the_fold",
	getSplitJsLoad: "_debug_split_js_load",
	getLogLevel: "_react_server_log_level",
	getLogLevelMain: "_react_server_log_level_main",
	getLogLevelTime: "_react_server_log_level_time",
	getLogLevelGauge: "_react_server_log_level_gauge"
};

var DebugUtil = {
	setRequest: function setRequest(req) {
		(0, _assign2.default)(RLS(), (0, _pick2.default)(req.getQuery(), (0, _map2.default)(DEBUG_PARAMS, function (v) {
			return v;
		})));
	},
	getAllDebugValues: function getAllDebugValues() {
		return (0, _assign2.default)({}, RLS());
	}
};

// Make the methods.
(0, _forEach2.default)(DEBUG_PARAMS, function (param, method) {
	return DebugUtil[method] = function () {
		return RLS()[param];
	};
});

exports.default = DebugUtil;
module.exports = exports["default"];
