// react-server/target/server/ReactServerAgent/Plugins.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Wrapper class around RLS-scoped ReactServerAgent plugins to avoid
 * a circular dependency between ReactServerAgent and ReactServerAgent/Request.
 *
 * Note: It's possible that this module shouldn't exist at all. ReactServerAgent
 * plugins maybe shouldn't be request-scoped; they could be set statically,
 * and get their data from RLS if they wanted, instead. i.e., store something
 * in RLS, then have the plugin retrieve it, and that way the code being
 * executed is the same for _all_ requests)
 */

var RLS = require('../util/RequestLocalStorage').getNamespace();

// Simple wrapper around an array the implements the
// API we want for adding/getting plugins

var PluginsHolder = function () {
	function PluginsHolder() {
		_classCallCheck(this, PluginsHolder);

		this.plugins = [];
	}

	_createClass(PluginsHolder, [{
		key: "asArray",
		value: function asArray() {
			// return a copy of the plugins array so that
			// the returned array can't change underneath
			// the caller
			return [].concat(this.plugins);
		}
	}, {
		key: "add",
		value: function add(plugin) {
			this.plugins.push(plugin);
		}
	}]);

	return PluginsHolder;
}();

function getPlugins(pluginType) {
	return RLS()[pluginType] || (RLS()[pluginType] = new PluginsHolder());
}

module.exports = {
	forRequest: getPlugins.bind(null, "request"),
	forResponse: getPlugins.bind(null, "response")
};
