// react-server/target/server/context/RequestContext.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Navigator = require('./Navigator'),
    RequestLocals = require('../util/RequestLocalStorage').getNamespace();

var RequestContext = function () {
	function RequestContext(routes) {
		var _this = this;

		_classCallCheck(this, RequestContext);

		this.navigator = new Navigator(this, routes);

		this.navigator.on('page', function (page) {
			_this.page = page;
		});

		this._navigateListeners = [];

		this.registerRequestLocal();
	}

	_createClass(RequestContext, [{
		key: 'registerRequestLocal',


		// when we transition from page to page, we clear out the RequestLocals, but we need to re-register
		// the RequestContext in the RequestLocals.
		value: function registerRequestLocal() {
			RequestLocals().instance = this;
		}
	}, {
		key: 'setServerStash',
		value: function setServerStash(stash) {
			this.serverStash = stash;
			return this;
		}
	}, {
		key: 'getServerStash',
		value: function getServerStash() {
			return this.serverStash;
		}
	}, {
		key: 'setDeviceType',
		value: function setDeviceType(type) {
			this.deviceType = type;
			return this;
		}
	}, {
		key: 'getDeviceType',
		value: function getDeviceType() {
			return this.deviceType;
		}
	}, {
		key: 'getCurrentPath',
		value: function getCurrentPath() {
			return this.navigator.getCurrentRoute().url;
		}
	}, {
		key: 'onNavigate',
		value: function onNavigate(callback) {
			this.navigator.on('navigateDone', callback);
		}
	}, {
		key: 'onNavigateStart',
		value: function onNavigateStart(callback) {
			this.navigator.on('navigateStart', callback);
		}
	}, {
		key: 'onLoadComplete',
		value: function onLoadComplete(callback) {
			this.navigator.on('loadComplete', callback);
		}
	}, {
		key: 'navigate',
		value: function navigate(request, type) {
			this.navigator.navigate(request, type);
		}
	}], [{
		key: 'getCurrentRequestContext',
		value: function getCurrentRequestContext() {
			return RequestLocals().instance;
		}
	}]);

	return RequestContext;
}();

var RequestContextBuilder = function () {
	function RequestContextBuilder() {
		_classCallCheck(this, RequestContextBuilder);
	}

	_createClass(RequestContextBuilder, [{
		key: 'setRoutes',
		value: function setRoutes(routes) {
			this.routes = routes;
			return this;
		}
	}, {
		key: 'setDefaultXhrHeadersFromRequest',
		value: function setDefaultXhrHeadersFromRequest() {
			return this;
		}
	}, {
		key: 'create',
		value: function create(extraOpts) {

			return new RequestContext(this.routes, extraOpts);
		}
	}]);

	return RequestContextBuilder;
}();

module.exports = RequestContext;
module.exports.Builder = RequestContextBuilder;
